lock file contains expected hash for pkg dependencies

ferd opened the pull request:
WIP that will let us store the hash for a package dependency in such a way that we can detect mismatches between the contents of an index and what is expected by the lock file.

This is expected to be crucial work when it comes to working with custom indexes today, or multiple indexes in the near future.

ferd commented:
Sample run:

```
λ rebar3 → pkg-local-hash-lock* → ./rebar3 escriptize
===> Verifying dependencies...
===> Fetching bbmustache ({pkg,<<"bbmustache">>,<<"1.0.4">>,
                                      <<"7BA94F971C5AFD7B6617918A4BB74705E36CAB36EB84B19B6A1B7EE06427AA38">>})
===> Downloaded package, caching at /home/ferd/.cache/rebar3/hex/default/packages/bbmustache-1.0.4.tar
===> Fetching certifi ({pkg,<<"certifi">>,<<"0.4.0">>,
                                   <<"A7965EFB868B179023618D29A407548F70C52466BF1849B9E8EBD0E34B7EA11F">>})
===> Downloaded package, caching at /home/ferd/.cache/rebar3/hex/default/packages/certifi-0.4.0.tar
===> The checksum for package at /home/ferd/.cache/rebar3/hex/default/packages/certifi-0.4.0.tar
(A7966EFB868B179023618D29A407548F70C52466BF1849B9E8EBD0E34B7EA11F) does
not match the checksum previously locked
(A7965EFB868B179023618D29A407548F70C52466BF1849B9E8EBD0E34B7EA11F). Either unlock or
upgrade the package, or make sure you fetched it from the same index from which it was initially fetched.

λ rebar3 → pkg-local-hash-lock* → ./rebar3 unlock certifi
λ rebar3 → pkg-local-hash-lock* → ./rebar3 escriptize
...
===> Compiling rebar
===> Building escript...
```

ferd commented:
Things this probably need:
- special clause for packages not to output the checksum on each fetch (that's verbose!)
- figuring out a test strategy for: no-hash to hash expected, broken hash and correct errors

I also figure we accept that the check only takes place on fetch, and not on subsequent runs since we don't store hashes and recomputing them is equivalent to repackaging all pkg deps on each run.

ferd commented:
Things I haven't conisdered: other untested providers (i.e. `tree`, `deps`), breakage of plugins that look at dependencies.

tsloughter commented:
Agreed it only take place on fetch.

Does this just skip the check if there isn't a checksum for the package in the lock file?

ferd commented:
If there isn't one, it will have the value `undefined` and the next time `install_deps` is run the current value will be taken and added to the lock file.

For example, I had nothing to do for rebar3's own lock file; just running code automatically plugged the hash in and updated the file.

talentdeficit commented on src/rebar_config.erl:
```diff
@@ -56,35 +56,92 @@ consult_lock_file(File) ->
-                  "upgrade Rebar3.", []),
+            case Vsn of
+                ?CONFIG_VERSION ->
+                    ok;
```
this is going to display the `newer version` warning with earlier lockfile versions, no?

ferd replied:
There are no earlier lock file versions (only versionless lock files) and there's a match on the format to match only those with a version in there.

ferd commented:
Sample run from a master rebar with this branch's lock file at a higher version:
```
λ rebar3 → pkg-local-hash-lock* → git co master
Switched to branch 'master'
Your branch is behind 'origin/master' by 2 commits, and can be fast-forwarded.
  (use "git pull" to update your local branch)
λ rebar3 → master* → cat rebar.lock
{"1.1.0",
[{<<"bbmustache">>,{pkg,<<"bbmustache">>,<<"1.0.4">>},0},
 ...
 {<<"ssl_verify_hostname">>,{pkg,<<"ssl_verify_hostname">>,<<"1.0.5">>},0}]}.
[
{pkg_hash,[
 {<<"bbmustache">>, <<"7BA94F971C5AFD7B6617918A4BB74705E36CAB36EB84B19B6A1B7EE06427AA38">>},
...
 {<<"ssl_verify_hostname">>, <<"2E73E068CD6393526F9FA6D399353D7C9477D6886BA005F323B592D389FB47BE">>}]}
].
λ rebar3 : master* → ./rebar3 escriptize && cp _build/default/bin/rebar3 rebar3 && chmod +x rebar3
===> Verifying dependencies...
===> Compiling rebar
===> Building escript...
λ rebar3 → master* → ./rebar3 compile
===> Rebar3 detected a lock file from a newer version. It will be loaded in compatibility mode, but important information may be missing or lost. It is recommended to upgrade Re
bar3.
===> Rebar3 detected a lock file from a newer version. It will be loaded in compatibility mode, but important information may be missing or lost. It is recommended to upgrade Re
bar3.
===> Rebar3 detected a lock file from a newer version. It will be loaded in compatibility mode, but important information may be missing or lost. It is recommended to upgrade Re
bar3.
===> Rebar3 detected a lock file from a newer version. It will be loaded in compatibility mode, but important information may be missing or lost. It is recommended to upgrade Re
bar3.
===> Verifying dependencies...
===> Compiling rebar
```
and then the lock file is reverted.

I guess it sucks that we display the warning on each consult, but we can't fix old versions now.

ferd commented:
Fixed the warning, added comment.

tsloughter commented:
+1
